### ISSUE TITLE
Support server and client testing

### ISSUE THREAD

**Issue by llxia** [adoptium/aqa-tests#2841]
In some cases, two machines are needed for testing. For example, some performance tests will need to communicate between server and client. We need to update our test pipeline to support this case.

Proposed solution:
- create a Jenkins job `Machine-Reservation` - this is a simple job to grep user-specified machine and sleep for the user-specified time. The main idea is to reserve the machine.
- in the test pipeline, a new parameter `RELATED_NODES` should be added. If the user provides a node name in `RELATED_NODES`, the test pipeline will trigger `Machine-Reservation` to book the specific machine before starting any testing. Otherwise, proceed with testing normally. For this case, we just need to pass in a single node. This can be further extended later for multiple nodes (i.e., comma separated).
- `RELATED_NODES` is set as an env variable to pass into the test scripts
- individual job will connect to `RELATED_NODES` for testing 
- once the testing is completed, abort `Machine-Reservation` if it is still running

**Comment by awsafsakif** [comment]
We also need a check to see if `CLIENT_NODE` machine is idle before triggering the reservation job. Otherwise we will run into an issue where if the client machine is already busy running a different benchmark, the reservation job will be scheduled but not triggered.

Extra variables required: TIMEOUT: How long server machine will be active in jenkins till job is cancelled, INTERVAL: How often to check if CLIENT_NODE is idle before triggering machine reservation and continuing with the benchmark run.

**Comment by llxia** [comment]
Yes, for checking if a node is available https://stackoverflow.com/questions/37227562/how-to-know-whether-the-jenkins-build-executors-are-idle-or-not-in-jenkins-using/37238045

Also, we have similar code to use the API: https://github.com/adoptium/aqa-tests/blob/master/buildenv/jenkins/openjdk_tests#L300

**Comment by smlambert** [comment]
Created https://ci.adoptopenjdk.net/job/Machine-Reservation job as a placeholder for a pipeline described in this issue.

@norbline - if you have not already done so, please rebase your branch (described in [Contributing.md](https://github.com/adoptium/aqa-tests/blob/master/Contributing.md)) in preparation to begin work on this issue.  I believe @llxia will share some code with you relating to Machine-Reservation shortly.

**Comment by llxia** [comment]
TODO:

- [ ] create Jenkins job Machine-Reservation https://ci.adoptopenjdk.net/view/Test_grinder/job/Machine-Reservation/

- [ ]  checkin initial Machine-Reservation script in aqa-tests repo (Lan)

- [ ] For now, If RELATED_NODES is provided and TARGET=IdleMicrobenchmark_HS, trigger Machine-Reservation job with RELATED_NODES and TIME_LIMIT

- [ ] Update IdleMicrobenchmark_HS to echo RELATED_NODES https://github.com/adoptium/aqa-tests/blob/master/perf/idle_micro/playlist.xml#L17

- [ ] get Machine-Reservation job status (job number). 
 - If the RELATED_NODES cannot be reserved within 30 mins, abort Machine-Reservation job and abort the job.
 - If the RELATED_NODES can be reserved within 30 mins, proceed to execute the test.
 
- [ ] in post-stage, release RELATED_NODES (abort Machine-Reservation job) and print the message

**Comment by smlambert** [comment]
Hi @norbline - can you add a comment in this issue, so I can assign it to you?  

Grinder_Perforama job is created for you to use to help test your work: https://ci.adoptopenjdk.net/job/Grinder_Perforama/

When you "Build with parameters, you should see the new parameter RELATED_NODES at the bottom just above the Build button.

**Comment by norbline** [comment]
Thanks, Shelley. All good to get on

**Comment by smlambert** [comment]
<img width="1658" alt="Screen Shot 2022-01-28 at 7 40 50 PM" src="https://user-images.githubusercontent.com/2836948/151639674-c92da0c3-5367-4aa3-9db9-6f1890da9dba.png">

**Comment by llxia** [comment]
@norbline For using Jenkins API to get job stage status, I found a similar thing at StackOverflow: https://stackoverflow.com/questions/37009906/access-stage-results-in-workflow-pipeline-plugin
Could you give it a try? Thanks

**Comment by norbline** [comment]
yes @Lan i will give it a try

**Comment by llxia** [comment]
After the investigation and experiments, we decided to use nested nodes instead of Jenkins API. In this way, we do not need to control the workflow/communication between two jobs. It is easy to maintain and less error-prone. 

The key changes for nested notes are delivered via #3364 and #3379.
Enhancements:

- [ ] check if we can get RELATED_NODES in playlist.xml. If not, set RELATED_NODES as env variable
- [ ] add `timestamp{}` to provide more info https://github.com/adoptium/aqa-tests/blob/25c804602873d503ec596371875c33a49fa88ad2/buildenv/jenkins/openjdk_tests#L100-L236
- [ ] if we can have `timestamp{}` that covers all the sections, we can remove individual timestamp{} in each stage.

**Comment by norbline** [comment]
Thanks @llxia

**Comment by llxia** [comment]
@awsafsakif could you test your case and let us know? Thanks.

**Comment by llxia** [comment]
Verified in [Grinder](https://ci.adoptopenjdk.net/job/Grinder_Perforama/91/console) that we can directly get `RELATED_NODES` in playlist. There is no need to set it as an env variable. 

In the console output, `test-docker-ubuntu2010-x64-2` is printed.

```
00:02:37.449  ===============================================
00:02:37.449  Running test IdleMicrobenchmark_HS_0 ...
00:02:37.449  ===============================================
00:02:37.449  IdleMicrobenchmark_HS_0 Start Time: Fri Mar 11 20:12:57 2022 Epoch Time (ms): 1647029577279
00:02:37.449  Nothing to be done for setup.
00:02:37.449  variation: -Xmx1024m
00:02:37.449  JVM_OPTIONS:  -Xmx1024m 
00:02:37.449  { itercnt=1; \
00:02:37.449  mkdir -p "/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../TKG/output_16470295765006/IdleMicrobenchmark_HS_0"; \
00:02:37.450  cd "/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../TKG/output_16470295765006/IdleMicrobenchmark_HS_0"; \
00:02:37.450  echo test-docker-ubuntu2010-x64-2; mkdir -p "/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../TKG/output_16470295765006/IdleMicrobenchmark_HS_0"; \
00:02:37.450  cd /home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../../jvmtest/perf/idle_micro; \
00:02:37.450  "/home/jenkins/workspace/Grinder_Perforama/openjdkbinary/j2sdk-image/bin/java"  -Xmx1024m  \
00:02:37.450  -cp ":/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../../jvmtest/perf/idle_micro/IdleMicrobenchmark.jar" net.adoptopenjdk.test.gc.idlemicrobenchmark.IdleMicroBenchmark \
00:02:37.450  --memoryLimit=700 --time=60 --idletime=300 --warmup=1 >"/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../TKG/output_16470295765006/IdleMicrobenchmark_HS_0"/std.out 2>&1 ; \
00:02:37.450  if [ $? -eq 0 ] ; then echo ""; echo "IdleMicrobenchmark_HS_0""_PASSED"; echo ""; cd /home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/..;  else echo ""; echo "IdleMicrobenchmark_HS_0""_FAILED"; echo ""; fi; } 2>&1 | tee -a "/home/jenkins/workspace/Grinder_Perforama/aqa-tests/TKG/../TKG/output_16470295765006/TestTargetResult";
00:02:37.450  test-docker-ubuntu2010-x64-2
```

FYI @awsafsakif

**Comment by llxia** [comment]
The feature is implemented. The task about - add timestamp{} to provide more info is optional and it is tacked in  https://github.com/adoptium/aqa-tests/issues/3396
Close this issue.
